### ISSUE TITLE
ssh does work, but sftp not working

### ISSUE THREAD

**Issue by freebrowser1** [ish-app/ish#2075]
This is a really nice app, much like Termux in Android.

But there is one flaw: sftp in an SSH file browser does not work. Probably a 'standard' Linux/SSH issue.

I connected using Forklift on macOS which resulted in an error (screenshot)
<img width="634" alt="Screenshot 2023-04-05 at 22 27 07" src="https://user-images.githubusercontent.com/26625141/230204365-0849cddd-9a17-4543-8099-92890ce3a575.png">
.

I changed sshd-config on the iSH server with the help of searching on this issue on internet by adding `PubkeyAuthentication yes` and  changing:
```
# override default of no subsystems
#Subsystem      sftp    /usr/lib/ssh/sftp-server
Subsystem       sftp    internal-sftp
```
but to no avail. Obviously I restarted iSH by closing and reopening it on the iPad itself by pressing Ctrl+D and then starting again which automatically starts sshd as I can immediately ssh into iSH again from my Mac.

iPadOS version 16.4

**Comment by wuebbel** [comment]
Same here. The problem is, I believe, that the sftp-server dies immediately after being called. If you call the sftp-server manually, you get:
```
# /usr/lib/ssh/sftp-server -e
unable to make the process undumpable
```
...and the sftp server exits immediately. If I get it right, then the relevant prctl is not available. One way out would probably be compiling a suitable sftp server, by changing in sftp_server.c
`        platform_disable_tracing(1);    /* strict */
`
to
`         platform_disable_tracing(0);    /* not strict */
`
Unfortunately, the choice strict/not strict cannot be made with ssh options.

**Comment by wuebbel** [comment]
Yes. this works.
```
$ scp out wuebbel@ipad-von-frank-2:
out                                           100%  385KB   1.6MB/s   00:00
```
The only thing I changed was

```
iPad-von-Frank-2:~/openssh-portable# grep platform_disable sftp-server.c
        platform_disable_tracing(0);    /* strict */

```

**Comment by freebrowser1** [comment]
@wuebbel: what did you change ?
You used an sftp client 'openssh-portable' ?

Does it have to do with that sftp-server uses SHA-1 instead of the usual SHA-2 RSA keys ?
If yes, can there be a SHA-2 variant which complies with most usual clients ?

**Comment by wuebbel** [comment]
@freebrowser1 No. I think you are completely on the wrong track. As I wrote above, call the sftp-sever executable directly. If it returns immediately, *that* is your problem. In this case, you have the option of either convincing the ish developers not to throw an error when an executable tries to protect itself, or compile your own sftp-server. To do the latter, you need to pull the original ssh sources, compile, install, and add the newly installed sftp server to sshd_config.
Something along the lines of (from my .ash_history)
```
apk install openssh-sftp-server gcc autoconf git automake musl-dev zlib-dev openssl openssl-dev make
git config --global pack.threads "1"
git clone --depth 10 https://github.com/openssh/openssh-portable.git
cd openssh/portable
vi sftp-server.c
./configure
make
make install
vi /etc/ssh/sshd_config
```

**Comment by freebrowser1** [comment]
Thanks for your effort, but it did not work.

```
apk add openssh-sftp-server gcc autoconf git automake musl-dev zlib-dev openssl openssl-dev make
git config --global pack.threads "1"
git clone --depth 10 https://github.com/openssh/openssh-portable.git
cd openssh/portable
```
Appeared to be **openssh-portable**
```
vi sftp-server.c
```
What should be changed here ?
```
./configure
```
This does not exist but there is a `./configure.ac`.
```
make
make install
vi /etc/ssh/sshd_config
```

**Comment by wuebbel** [comment]
@freebrowser1 Change the line I mentioned in the first post.

> by changing in sftp_server.c

`       platform_disable_tracing(1);    /* strict */

> to

`       platform_disable_tracing(0);    /* not strict */`

I missed an "autoreconf" in my .ash_history. You'd better follow the instructions in the git, though. I'll add the autoreconf to my instructions.

**Comment by freebrowser1** [comment]
@wuebbel 
Thanks. It compiled until `make install` which did compile several files but errored out at:

```

_SOURCE -D_GNU_SOURCE -DSSHDIR=\"/usr/local/etc\" -D_PATH_SSH_PROGRAM=\"/usr/local/bin/ssh\" -D_PATH_SSH_ASKPASS_DEFAULT=\"/usr/local/libexec/ssh-askpass\" -D_PATH_SFTP_SERVER=\"/usr/local/libexec/sftp-server\" -D_PATH_SSH_KEY_SIGN=\"/usr/local/libexec/ssh-keysign\" -D_PATH_SSH_PKCS11_HELPER=\"/usr/local/libexec/ssh-pkcs11-helper\" -D_PATH_SSH_SK_HELPER=\"/usr/local/libexec/ssh-sk-helper\" -D_PATH_SSH_PIDDIR=\"/var/run\" -D_PATH_PRIVSEP_CHROOT_DIR=\"/var/empty\" -DHAVE_CONFIG_H -c ssh-sk-client.c -o ssh-sk-client.o
cc -o ssh ssh.o readconf.o clientloop.o sshtty.o sshconnect.o sshconnect2.o mux.o ssh-sk-client.o -L. -Lopenbsd-compat/  -Wl,-z,relro -Wl,-z,now -Wl,-z,noexecstack -fstack-protector-strong -pie  -lssh -lopenbsd-compat   -lcrypto  -lz
/usr/lib/gcc/i586-alpine-linux-musl/10.3.1/../../../../i586-alpine-linux-musl/bin/ld: ./libssh.a(sshkey.o):(.data.rel.ro+0x3c): undefined reference to `sshkey_rsa_impl'
/usr/lib/gcc/i586-alpine-linux-musl/10.3.1/../../../../i586-alpine-linux-musl/bin/ld: ./libssh.a(sshkey.o):(.data.rel.ro+0x40): undefined reference to `sshkey_rsa_cert_impl'
/usr/lib/gcc/i586-alpine-linux-musl/10.3.1/../../../../i586-alpine-linux-musl/bin/ld: ./libssh.a(sshkey.o):(.data.rel.ro+0x44): undefined reference to `sshkey_rsa_sha256_impl'
/usr/lib/gcc/i586-alpine-linux-musl/10.3.1/../../../../i586-alpine-linux-musl/bin/ld: ./libssh.a(sshkey.o):(.data.rel.ro+0x48): undefined reference to `sshkey_rsa_sha256_cert_impl'
/usr/lib/gcc/i586-alpine-linux-musl/10.3.1/../../../../i586-alpine-linux-musl/bin/ld: ./libssh.a(sshkey.o):(.data.rel.ro+0x4c): undefined reference to `sshkey_rsa_sha512_impl'
/usr/lib/gcc/i586-alpine-linux-musl/10.3.1/../../../../i586-alpine-linux-musl/bin/ld: ./libssh.a(sshkey.o):(.data.rel.ro+0x50): undefined reference to `sshkey_rsa_sha512_cert_impl'
collect2: error: ld returned 1 exit status
make: *** [Makefile:207: ssh] Error 1
```
BTW : What does i586 mean ? The iPad runs arm64 and not Intel ?

**Comment by khurshid-alam** [comment]
@freebrowser1 Did you file a bug in OpenSSH-portable ? What happens when you use stable branch instead master?

**Comment by mikeggh** [comment]
I do not believe that alone would fix it.  Here is a simple LD_PRELOAD library which essentially makes that function irrelevant.

in test.c:
```
#include <sys/prctl.h>

int prctl(int op, ...
                 /* unsigned long arg2, unsigned long arg3,
                 unsigned long arg4, unsigned long arg5 */ ){
return 0;
}
```
compile: gcc -c test.c -fPIC -shared -o test.so

prepare:
export LD_PRELOAD=/root/test.so 
rc-service sshd reload

still crash :<

**Comment by umarbarrister** [comment]
> [@wuebbel](https://github.com/wuebbel): what did you change ? You used an sftp client 'openssh-portable' ?
> 
> Does it have to do with that sftp-server uses SHA-1 instead of the usual SHA-2 RSA keys ? If yes, can there be a SHA-2 variant which complies with most usual clients ?

Preparing to unpack .../command-not-found_3.4.1-7_aarch64.deb ...
Unpacking command-not-found (3.4.1-7) over (3.4.1-6) ...
Preparing to unpack .../libexpat_2.8.0_aarch64.deb ...
Unpacking libexpat (2.8.0) over (2.7.5) ...
Preparing to unpack .../git_2.54.0_aarch64.deb ...
Unpacking git (2.54.0) over (2.53.0) ...
Preparing to unpack .../openssh-sftp-server_10.3p1-1_aarch64.deb
